File: app/src/main/java/com/balicodes/quicksms/SMSListFragment.java
package com.balicodes.quicksms;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.media.MediaPlayer;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.FrameLayout;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by eka on 6/27/15.
 */
public class SMSListFragment extends Fragment implements AdapterView.OnItemClickListener {
    private final String TAG = SMSListFragment.class.toString();

    private DBHelper dbHelper;
    private List<SMSItem> listSMS = new ArrayList<SMSItem>();
    private SMSListAdapter listAdapter;
    private ListView listView;
    private FrameLayout tapInfo;
    private boolean afterSending = false;
    private MediaPlayer beep;
    private SentReceiver sentReceiver;
    private DeliveryReceiver deliveryReceiver;
    private boolean isReceiverRegistered = false;
    private SMSItem currentSMSitem;
    private int currentSMSitemIndex;
    private int currentSendingCount = 0;
    private SharedPreferences sp;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        dbHelper = new DBHelper(getActivity());
        listAdapter = new SMSListAdapter(getActivity(), listSMS);
        beep = MediaPlayer.create(getActivity(), R.raw.beep);
        beep.setVolume((float) 0.5, (float) 0.5);
    }

    private void loadList() {
        listSMS.clear();
        List<SMSItem> all = dbHelper.all();

        // show hide tap info frame.
        if (all.size() == 0) {
            tapInfo.setVisibility(View.GONE);
        } else {
            tapInfo.setVisibility(View.VISIBLE);
        }

        for (SMSItem item : all) {
            listSMS.add(item);
        }
        listAdapter.notifyDataSetChanged();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.sms_list_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        switch (id) {
            case R.id.action_new:
                FragmentTransaction ft = getActivity().getSupportFragmentManager().beginTransaction();
                ft.replace(R.id.container, new SMSFormFragment());
                ft.addToBackStack(null);
                ft.commit();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.sms_list_fragment, container, false);
        tapInfo = (FrameLayout) view.findViewById(R.id.tap_info);
        listView = (ListView) view.findViewById(R.id.listView);
        listView.setAdapter(listAdapter);
        listView.setOnItemClickListener(this);
        registerForContextMenu(listView);

        return view;
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        if (afterSending) {
            afterSending = false;
            return;
        }

        SMSItem smsitem = listSMS.get(position);
        SMSFormFragment form = new SMSFormFragment();
        form.setArguments(smsitem.toBundle());

        getActivity().getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.container, form)
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                .addToBackStack("sms_form")
                .commit();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        ((MainActivity) getActivity()).getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        loadList();
    }

    /*----------------------------------------------------------------------------------------------
    Try to send sms on long tap.
    ----------------------------------------------------------------------------------------------*/
    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {

        if (currentSMSitem != null) {
            resetCurrentSMSItem();
        }

        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuInfo;
        currentSMSitem = listSMS.get(info.position);
        currentSMSitemIndex = info.position;

        tryToSend();
    }

    /*----------------------------------------------------------------------------------------------
    Register sending and delivery receivers.
    ----------------------------------------------------------------------------------------------*/
    private void registerReceivers() {
        sentReceiver = new SentReceiver();
        deliveryReceiver = new DeliveryReceiver();

        getActivity().registerReceiver(sentReceiver, new IntentFilter(Config.SENT_STATUS_ACTION));
        getActivity().registerReceiver(deliveryReceiver, new IntentFilter(Config.DELIVERY_STATUS_ACTION));

        isReceiverRegistered = true;
        Log.d(TAG, "Sent and Delivery receivers registered.");
    }

    /*----------------------------------------------------------------------------------------------
    Un-register sending and delivery receivers.
    ----------------------------------------------------------------------------------------------*/
    private void unregisterReceivers() {
        try {
            if (isReceiverRegistered) {
                getActivity().unregisterReceiver(sentReceiver);
                getActivity().unregisterReceiver(deliveryReceiver);
                Log.d(TAG, "Sent and Delivery receivers un-registered.");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    /* ---------------------------------------------------------------------------------------------
    Try to send the message and show confirmation dialog if user set so.
    ----------------------------------------------------------------------------------------------*/
    private void tryToSend() {
        afterSending = true;
        boolean confirm = sp.getBoolean(getString(R.string.pref_sending_confirmation_key), false);

        // If sending confirmation required.
        if (confirm) {
            AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
            builder.setTitle(R.string.confirm_sending);

            builder.setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    sendSMS();
                }
            });
            builder.setNegativeButton(R.string.no, new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                }
            });
            AlertDialog dialog = builder.create();
            dialog.show();

        } else {
            sendSMS();
        }
    }

    /*----------------------------------------------------------------------------------------------
    The actual function to send the message.
     ---------------------------------------------------------------------------------------------*/
    private void sendSMS() {
        if (currentSMSitem == null) return;

        // Android API 23+ requires this
        if(Build.VERSION.SDK_INT == 26){
            int permissionSendSMS = ContextCompat.checkSelfPermission(getActivity(), android.Manifest.permission.SEND_SMS);
            int permissionReadPhoneState = ContextCompat.checkSelfPermission(getActivity(), Manifest.permission.READ_PHONE_STATE);

            if(permissionSendSMS != PackageManager.PERMISSION_GRANTED || permissionReadPhoneState != PackageManager.PERMISSION_GRANTED){
                String[] perms = new String[]{Manifest.permission.SEND_SMS, Manifest.permission.READ_PHONE_STATE};
                ActivityCompat.requestPermissions(getActivity(), perms, Config.SEND_SMS_PERMISSION_REQUEST);
                return;
            }
        }else{
            int permission = ContextCompat.checkSelfPermission(getActivity(), android.Manifest.permission.SEND_SMS);
            if(permission != PackageManager.PERMISSION_GRANTED){
                String[] perms = new String[]{Manifest.permission.SEND_SMS};
                ActivityCompat.requestPermissions(getActivity(), perms, Config.SEND_SMS_PERMISSION_REQUEST);
                return;
            }
        }

        Log.d(TAG, "Sending sms: " + currentSMSitem.getTitle());
        listAdapter.setSending(currentSMSitemIndex);
        boolean playBeep = sp.getBoolean(getString(R.string.pref_enable_beep_key), true);
        if (playBeep) beep.start();

        // start our sending service
        Intent sendingIntent = new Intent(getActivity().getApplicationContext(), SendingService.class);
        sendingIntent.putExtra(Config.SMS_BUNDLE_EXTRA_KEY, currentSMSitem.toBundle());
        getActivity().startService(sendingIntent);
    }

    private void resendSingleRecipient() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(R.string.sending_failed);
        builder.setMessage(R.string.sending_failed_info);
        builder.setPositiveButton(R.string.retry, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int id) {
                currentSendingCount = 0;
                sendSMS();
            }
        });
        builder.setNegativeButton(R.string.ok, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int id) {
                resetCurrentSMSItem();
            }
        });
        AlertDialog dialog = builder.create();
        dialog.show();
    }

    /*----------------------------------------------------------------------------------------------
    Private class to handle Sending status.
    ----------------------------------------------------------------------------------------------*/
    private class SentReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            Bundle recBundle = intent.getBundleExtra(Config.RECIPIENT_EXTRA_KEY);

            if (getResultCode() == Activity.RESULT_OK) {
                if (recBundle != null) {
                    Recipient rec = Recipient.fromBundle(recBundle);
                    listAdapter.notifyStatusChanged(currentSMSitemIndex, SMSItem.STATUS_SENT, rec);
                    Log.d("SentReceiver", "====> OK: " + rec.getNumber());
                }
            } else {
                if (recBundle != null) {
                    Recipient rec = Recipient.fromBundle(recBundle);
                    Log.d("SentReceiver", String.format("%s - %s - %s", rec.getId(), rec.getName(), rec.getNumber()));

                    listAdapter.notifyStatusChanged(currentSMSitemIndex, SMSItem.STATUS_FAILED, rec);
                    Log.d("SentReceiver", "====> FAILED: " + rec.getNumber());
                }
            }

            // If finish sending, check whether all sent successfully.
            // if not, show resend screen with all failed recipients.
            currentSendingCount++;
            if (currentSendingCount == currentSMSitem.totalRecipients()) {

                if (currentSMSitem.failedInfoList.size() > 0) {
                    afterSending = true;

                    // If the recipient only one, no need to show Resend screen
                    if (currentSMSitem.totalRecipients() == 1) {
                        resendSingleRecipient();
                    } else {

                        boolean confirm_before_send = sp.getBoolean(getString(R.string.pref_sending_confirmation_key), false);
                        if (confirm_before_send) {
                            afterSending = false;
                        }

                        Intent resendIntent = new Intent(getActivity().getApplicationContext(), ResendActivity.class);
                        resendIntent.putExtra(Config.SMS_MESSAGE_EXTRA_KEY, currentSMSitem.getMessage());
                        resendIntent.putParcelableArrayListExtra(Config.RECIPIENT_PARCELS_EXTRA_KEY, currentSMSitem.failedInfoList);
                        startActivity(resendIntent);
                    }
                }
            }
        }
    }

    /*----------------------------------------------------------------------------------------------
    Private class to handle Delivery status.
    ----------------------------------------------------------------------------------------------*/
    private class DeliveryReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {

            Bundle recBundle = intent.getBundleExtra(Config.RECIPIENT_EXTRA_KEY);

            if (getResultCode() == Activity.RESULT_OK) {
                if (recBundle != null) {
                    Recipient rec = Recipient.fromBundle(recBundle);
                    listAdapter.notifyStatusChanged(currentSMSitemIndex, SMSItem.STATUS_DELIVERED, rec);
                    Log.d("DeliveryReceiver", "====> OK: " + rec.getNumber());
                }
            }
        }
    }

    /*----------------------------------------------------------------------------------------------
    Reset current selected SMSItem state.
    ----------------------------------------------------------------------------------------------*/
    private void resetCurrentSMSItem() {
        currentSMSitem = null;
        currentSMSitemIndex = -1;
        currentSendingCount = 0;
        afterSending = false;
        Log.d(TAG, "Reset currentSMSItem");
    }

    @Override
    public void onPause() {
        super.onPause();
        if (dbHelper != null) {
            dbHelper.close();
        }
        unregisterReceivers();
    }

    @Override
    public void onStop() {
        super.onStop();
        if (dbHelper != null) {
            dbHelper.close();
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (dbHelper != null) {
            dbHelper.close();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        registerReceivers();
        sp = PreferenceManager.getDefaultSharedPreferences(getActivity());

    }
}

File: app/src/main/java/com/balicodes/quicksms/SMSListFragment.kt
/*
 * Copyright (C) 2016-2018 Eka Putra
 *
 * Quick SMS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.balicodes.quicksms

import android.app.AlertDialog
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.util.Log
import android.view.*
import android.widget.AdapterView
import android.widget.FrameLayout
import android.widget.ListView
import com.balicodes.quicksms.entity.MessageEntity
import com.balicodes.quicksms.model.SMSItem
import com.balicodes.quicksms.service.SendingService
import com.balicodes.quicksms.util.SmsPermissionChecker
import com.balicodes.quicksms.viewmodel.MessageViewModel
import java.util.*
import java.util.logging.Logger

class SMSListFragment : Fragment(), AdapterView.OnItemClickListener {

    private val TAG = SMSListFragment::class.java.toString()

    private lateinit var viewModel: MessageViewModel
    private lateinit var tapInfo: FrameLayout
    private lateinit var sp: SharedPreferences

    private val listSMS = ArrayList<SMSItem>()
    private var listAdapter: SMSListAdapter? = null
    private var afterSending = false
    private var currentSMSitem: SMSItem? = null
    private var currentSMSitemIndex: Int = 0
    private var currentSendingCount = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        val sp = PreferenceManager.getDefaultSharedPreferences(requireContext())
        val sortBy = sp.getString(requireContext().getString(R.string.pref_sort_by_key), requireContext().getString(R.string.pref_sort_by_default_value))

        viewModel = ViewModelProviders.of(requireActivity()).get(MessageViewModel::class.java)
        viewModel.setOrderBy(sortBy)
        viewModel.getMessages().observe(this, Observer<List<MessageEntity>> {
            listSMS.clear()

            // show hide tap info frame.
            if (it!!.isEmpty()) {
                tapInfo.visibility = View.GONE
            } else {
                tapInfo.visibility = View.VISIBLE
            }

            val items = ArrayList<SMSItem>()
            it.forEach { items.add(it.toSmsItem()) }

            listSMS.addAll(items)
            listAdapter!!.notifyDataSetChanged()
        })

        listAdapter = SMSListAdapter(listSMS)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater!!.inflate(R.menu.sms_list_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        val id = item!!.itemId

        when (id) {
            R.id.action_new -> {
                viewModel.selectMessage(null)

                val ft = requireActivity().supportFragmentManager.beginTransaction()
                ft.replace(R.id.container, SMSFormFragment())
                ft.addToBackStack(null)
                ft.commit()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.sms_list_fragment, container, false)

        tapInfo = view.findViewById(R.id.tap_info)

        val listView = view.findViewById<ListView>(R.id.listView)
        listView.adapter = listAdapter
        listView.onItemClickListener = this

        registerForContextMenu(listView)

        return view
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        if (afterSending) {
            afterSending = false
            return
        }

        val item = listSMS[position]
        viewModel.selectMessage(item.toEntity())

        val form = SMSFormFragment()
        requireActivity().supportFragmentManager
                .beginTransaction()
                .replace(R.id.container, form)
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                .addToBackStack("sms_form")
                .commit()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        (requireActivity() as MainActivity).supportActionBar!!.setDisplayHomeAsUpEnabled(false)
    }

    /*----------------------------------------------------------------------------------------------
    Try to send sms on long tap.
    ----------------------------------------------------------------------------------------------*/
    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo) {

        if (currentSMSitem != null) {
            resetCurrentSMSItem()
        }

        val info = menuInfo as AdapterView.AdapterContextMenuInfo
        currentSMSitem = listSMS[info.position]
        currentSMSitemIndex = info.position

        tryToSend()
    }

    /* ---------------------------------------------------------------------------------------------
    Try to send the message and show confirmation dialog if user set so.
    ----------------------------------------------------------------------------------------------*/
    private fun tryToSend() {
        afterSending = true
        val confirm = sp.getBoolean(getString(R.string.pref_sending_confirmation_key), false)

        // If sending confirmation required.
        if (confirm) {
            val builder = AlertDialog.Builder(requireContext())
            builder.setTitle(R.string.confirm_sending)

            builder.setPositiveButton(R.string.yes) { dialog, id -> sendSMS() }
            builder.setNegativeButton(R.string.no) { dialog, id -> }
            val dialog = builder.create()
            dialog.show()

        } else {
            sendSMS()
        }
    }

    /*----------------------------------------------------------------------------------------------
    The actual function to send the message.
     ---------------------------------------------------------------------------------------------*/
    private fun sendSMS() {
        if (currentSMSitem == null) return

        // Check for permission to send sms.
        if (!SmsPermissionChecker.hasPermissionToSendSms(requireContext())) {
            LOG.warning("Permission not granted. Asking for permission...")

            SmsPermissionChecker.requestSendSmsPermission(requireActivity())
            return
        }

        // start our sending service
        val sendingIntent = Intent(requireContext(), SendingService::class.java)
        sendingIntent.putExtra(Config.SMS_BUNDLE_EXTRA_KEY, currentSMSitem!!.toBundle())
        requireActivity().startService(sendingIntent)
    }

    private fun resendSingleRecipient() {
        val builder = AlertDialog.Builder(requireContext())
        builder.setTitle(R.string.sending_failed)
        builder.setMessage(R.string.sending_failed_info)
        builder.setPositiveButton(R.string.retry) { dialog, id ->
            currentSendingCount = 0
            sendSMS()
        }
        builder.setNegativeButton(R.string.ok) { dialog, id -> resetCurrentSMSItem() }
        val dialog = builder.create()
        dialog.show()
    }

    /*----------------------------------------------------------------------------------------------
    Reset current selected SMSItem state.
    ----------------------------------------------------------------------------------------------*/
    private fun resetCurrentSMSItem() {
        currentSMSitem = null
        currentSMSitemIndex = -1
        currentSendingCount = 0
        afterSending = false
        Log.d(TAG, "Reset currentSMSItem")
    }

    override fun onResume() {
        super.onResume()
        sp = PreferenceManager.getDefaultSharedPreferences(requireContext())
    }

    companion object {
        private val LOG: Logger = Logger.getLogger(SMSListFragment::class.java.name)
    }
}
